### PR TITLE
Section cut views

### PR DESCRIPTION
**Pull request by pmconne**
DisplayStyleSettings now has a ClipStyle property consisting of:
- a boolean indicating whether cut geometry should be produced; and
- a CutStyle specifying how the cut geometry should be drawn (ViewFlags, edge symbology, and optional symbology overrides);

The section-cut graphics contain planar polyfaces for solid geometry (closed volumes); and, if edges are enabled for the view, planar line strings for sheets.

When drawing a view that has a clip and a clip style specifying cut graphics should be produced, we augment each design model's tile tree with another tile tree that produces only the cut geometry. The clip vector is encoded in compressed form into the tile tree's Id. During tile selection we trivially cull tiles *not* intersecting the clipping planes.

Incidentally cleaned up JSON types for clipping-related types (used to be `any`).

TODO for this PR:
- Test with various scenarios (animation, classification, etc etc).

TODO longer-term:
- Produce cut graphics for dynamic element graphics (interactive editing).
- Address ugly edges produced for cut geometry if original geometry was a polyface.
- Generate section-cut geometry for element graphics (interactive editing workflows)
- Somehow(?) produce section-cut graphics for geometry that is transformed by a schedule script??
- Allow limited symbology overrides of geometry inside or outside the clipping planes
  - This could only affect line pattern, color, and locatability; maybe an option to discard surfaces (only useful if view has edges enabled). Not hugely useful.

[Corresponding native PR](https://dev.azure.com/bentleycs/iModelTechnologies/_git/imodel02/pullrequest/130286).

### CODE REVIEW

**Comment by pmconne** [comment]
Closing - wrong target branch.

**Review by bbastings — COMMENTED** [review]
ViewClipModifyTool changes are approved.

**Review by markschlosseratbentley — CHANGES_REQUESTED** [review]

File: docs/changehistory/NextVersion.md [hunk 1]
```diff
@@ -3,6 +3,27 @@ ignore: true
 ---
 # NextVersion
 
+## Section-cut graphics
+
+A [DisplayStyleState]($frontend) can now be configured to produce section-cut graphics from the view's [ClipVector]($geometry-core). In the image below, a clipping plane has been applied to produce a cross-section view of a house. Note that all of the geometry intersecting the clipping plane appears to be solid, whereas in the real world we'd expect things like walls, floors, and tables to be solid throughout:
```
[the remark this answers is not in the document]
Comment (pmconne) [reply docs/changehistory/NextVersion.md]: /s/solid/hollow

[the remark this answers is not in the document]
Comment (markschlosseratbentley) [reply docs/changehistory/NextVersion.md]: Good stuff.  Another minor typo -- 'hoolow'

[the remark this answers is not in the document]
Comment (pmconne) [reply docs/changehistory/NextVersion.md]: What a dummy. Fixed.

**Comment by markschlosseratbentley** [comment]
Is it expected that clip style lines should look like this?

image

image

**Comment by pmconne** [comment]
> Is it expected that clip style lines should look like this?

@markschlosseratbentley Yes, that keyin is for testing overriding aspects of section-cut graphics symbology.
[Here's the code that produces it](https://github.com/imodeljs/imodeljs/blob/6f6bef8b1040ce375d6809f70b36dd4d8d39f485/core/frontend-devtools/src/tools/ClipTools.ts#L138).

**Comment by pmconne** [comment]
@markschlosseratbentley if your question is about the diagonal edges, that's mentioned in the TODO in PR description.

**Comment by markschlosseratbentley** [comment]
> @markschlosseratbentley if your question is about the diagonal edges, that's mentioned in the TODO in PR description.

Yes, that's what I was referring to - Ah, okay, I see it in the TODO.

**Comment by markschlosseratbentley** [comment]
> > Is it expected that clip style lines should look like this?
> 
> @markschlosseratbentley Yes, that keyin is for testing overriding aspects of section-cut graphics symbology.
> [Here's the code that produces it](https://github.com/imodeljs/imodeljs/blob/6f6bef8b1040ce375d6809f70b36dd4d8d39f485/core/frontend-devtools/src/tools/ClipTools.ts#L138).

Thanks.

**Comment by markschlosseratbentley** [comment]
I built this and tested it out - it works well for me.

**Review by markschlosseratbentley — APPROVED** [review]

**Comment by pmconne** [comment]
> ViewClipModifyTool changes are approved.

@bbastings but you didn't click "Approve".

**Review by bbastings — APPROVED** [review]

**Pull request merged**